src/QDL/qdlDevice.js

```diff
@@ -0,0 +1,181 @@
+import { usbClass } from "./usblib"
+import { Sahara } from  "./sahara"
+import { Firehose } from "./firehose"
+import { AB_FLAG_OFFSET, AB_PARTITION_ATTR_SLOT_ACTIVE } from "./gpt"
+import { concatUint8Array, runWithTimeout } from "./utils"
+
+
+export class qdlDevice {
+  constructor() {
+    this.mode             = "";
+    this.cdc              = new usbClass();
+    this.sahara           = new Sahara(this.cdc);
+    this.firehose         = new Firehose(this.cdc);
+    this._connectResolve  = null;
+    this._connectReject   = null;
+  }
+
+
+  async waitForConnect() {
+    return await new Promise((resolve, reject) => {
+      this._connectResolve = resolve;
+      this._connectReject  = reject;
+    });
+  }
+
+
+  async connectToSahara() {
+    while (!this.cdc.connected) {
+      await this.cdc?.connect();
+      if (this.cdc.connected) {
+        console.log("QDL device detected");
+        let resp = await runWithTimeout(this.sahara?.connect(), 10000);
+        if (resp.hasOwnProperty("mode")) {
+          this.mode = resp["mode"];
+          console.log("Mode detected:", this.mode);
+          return resp;
+        }
+      }
+    }
+    return {"mode" : "error"};
+  }
+
+
+  async flashBlob(partitionName, blob, onProgress=(_progress)=>{}) {
+    let startSector = 0;
+    let dp = await this.firehose?.detectPartition(partitionName);
+    const found = dp[0];
+    if (found) {
+      let lun = dp[1];
+      const imgSize = blob.size;
+      let imgSectors = Math.floor(imgSize/this.firehose.cfg.SECTOR_SIZE_IN_BYTES);
+      if (imgSize % this.firehose.cfg.SECTOR_SIZE_IN_BYTES !== 0)
+        imgSectors += 1;
+      if (partitionName.toLowerCase() !== "gpt") {
+        const partition = dp[2];
+        if (imgSectors > partition.sectors) {
+          console.error("partition has fewer sectors compared to the flashing image");
+          return false;
+        }
+        startSector = partition.sector;
+        console.log(`Flashing ${partitionName}...`);
+        if (await this.firehose.cmdProgram(lun, startSector, blob, (progress) => onProgress(progress))) {
+          console.log(`partition ${partitionName}: startSector ${partition.sector}, sectors ${partition.sectors}`);
+        } else {
+          throw new Error(`Errow while writing ${partitionName}`);
+        }
+      }
+    } else {
+      throw new Error(`Can't find partition ${partitionName}`);
+    }
+    return true;
+  }
+
+
+  async resetUserdata() {
```

Comment:
this isn't a generic QDL function - this is a quirk of how we want to do the erase and format